### PR TITLE
Add parsing of decimal constants (e.g., 1.02e+01)

### PR DESCRIPTION
**Pull request by proux01**
This adds parsing of decimal constant like `1.02e+01` (= `10.2`). More precisely, it accepts inputs following the regexp `[0-9]* [.]? [0-9]+ ([eE][+-]?[0-9]+)?`. Parsers are provided for Reals and Q. This may look like a crazy useless feature but we'll sorely need it in a future PR adding primitive floating point computation (c.f. #8276 with @erikmd).

Note that we request at least one digit after a potential dot to avoid things like `Check 42.` to be parsed as "Check" "42." instead of "Check" "42" ".".

Note : the 'is_number : string -> bool' function was somewhat duplicated in the code, it is now factored in 'lib/util.ml', not sure it's the best place though.

Note : Should `IZR 10200 / IZR (Z.pow_pos 10 3)` (resp. `Qmake 10200 1000`) be uninterpreted as `10200e-3` or `102e-1` or `1.02e1`?

Todo:
- [x] separate numerals and notations? (c.f. https://github.com/coq/coq/pull/8764#discussion_r251377776)
- [x] accept `_` as comment in numerals ?
- [x] add hexadecimal/octal ? (in another PR)
- [x] check that change under notations works (c.f. https://github.com/coq/coq/pull/8764#issuecomment-458084739)
- [x] require at least one digit before dot to avoid conflict with pair projectors `.1` and `.2`
- [x] use a ring/micromega friendly encoding
- [x] update documentation
- [x] add overlays

~~Depends on: #8720~~ (merged)
~~Depends on: #9815~~ (merged)

Overlays:
- https://github.com/coq-community/corn/pull/72
- https://github.com/HoTT/HoTT/pull/1022
- https://github.com/ppedrot/ltac2/pull/116
- https://github.com/QuickChick/QuickChick/pull/150
- https://github.com/coq/stdlib2/pull/18

### CODE REVIEW

File: interp/constrexpr.ml
Comment (herbelin) [inline interp/constrexpr.ml]: More generally, let me make explicit that refining half-way the structure of `Numeral` as you do is somehow strange, compared to do the decomposition in the `coqdecimal_of_rawnum` function. But why not.

**Review by herbelin — COMMENTED** [review]
Looks very good to me. Thanks a lot for this contribution.

See comments in the code. Otherwise, here are some answers and general comments.

> Note : the 'is_number : string -> bool' function was somewhat duplicated in the code, it is now factored in 'lib/util.ml', not sure it's the best place though.

No opinion, maybe @ejgallego has a suggestion.

> Note : Should `IZR 10200 / IZR (Z.pow_pos 10 3)` (resp. `Qmake 10200 1000`) be uninterpreted as `10200e-3` or `102e-1` or `1.02e1`?

As a naive user, I would probably expect `10.2` if in a type with a canonical form. If in `Q`, without normalization, I guess I would expect to see `10200/1000`. And for `IZR 10200 / IZR (Z.pow_pos 10 3)`, I would expect, I believe,`10200 / 10^3`.

> Note that we request at least one digit after a potential dot to avoid things like Check 42. to be parsed as "Check" "42." instead of "Check" "42" ".".

Yes, but this shall be mainly a lexer issue, I guess.

As far as I understand, this is only the interpretation infrastructure. It would remain to add the lexer, as well as providing effective primitive interpreters and/or extending the user-defined numeral notations with floating point numbers, right?

More generally, this raises the question of other kinds of printers. E.g. parsing `0xFF00`. Also, aren't there other standard than using `e` for floating numbers. The `e` looks pretty programming-language (confusing) jargon.

**Review by ejgallego — COMMENTED** [review]
Umm, that seems like a big patch for a fairly small feature. I wonder if the lexer shouldn't just process the literals away.

**Comment by silene** [comment]
> Maybe @silene would have other ideas?

Not really. The `e` notation has been in use for more than 60 years and I guess it has become a de facto standard. The only modern tool I can think of that does not use it is Mathematica, which instead uses `123*^-1`, because `123e-1` is parsed as `123 * e - 1`.

**Comment by proux01** [comment]
@herbelin all your comments are addressed in the latest rebase. More specifically:
- the half-way refinement of Numeral has been dropped and the decomposition is now entirely done in `coqdecimal_of_rawnum`
- no new token, the "INT" token is reused, and renamed "NUMERAL" to avoid any confusion
- no space after "-" sign
- `type sign = bool` has been replaced by `type sign = SPlus | SMinus`

**Comment by silene** [comment]
> Maybe @silene would have other ideas?

Not really. The `e` notation has been in use for more than 60 years and I guess it has become a de facto standard. The only modern tool I can think of that does not use it is Mathematica, which instead uses `123*^-1`, because `123e-1` is parsed as `123 * e - 1`.

**Comment by proux01** [comment]
That was indeed my first thought, however it appeared a non issue in our developments mixing reals and pair projectors. Notations such as `p .1` and `p .2` are interpreted before numeral notations and everything works fine. Mathcomp perfectly compiles and one can even write:

```
Check ((2.54, .2).1)%R.
(254e-2, 2e-1).1 : R
```

File: parsing/tok.ml
Comment (ejgallego) [inline parsing/tok.ml]: I would strongly prefer if this was a properly typed record with the components of the numeral token.

**Review by ejgallego — DISMISSED** [review]

**Comment by ggonthier** [comment]
> That was indeed my first thought, however it appeared a non issue in our developments mixing reals and pair projectors. Notations such as `p .1` and `p .2` are interpreted before numeral notations and everything works fine. Mathcomp perfectly compiles and one can even write:
> 
> ```
> Check ((2.54, .2).1)%R.
> (254e-2, 2e-1).1 : R
> ```

Certainly, but it's the other way around: when using mathcomp one won't be able to use `.1` to denote a decimal (this will likely cause an obscure typing error), and, depending on how the lexer has been modified, `.1414` might also be broken, while `.5` and `.33` will continue to work. So to provide consistent behavior the ssreflect plugin would have to modify the lexer to enforce a leading digit before decimals, and the manual would have to document this restriction. The ssreflect maintainers would probably prefer  this to be the default behavior.

File: parsing/tok.ml
Comment (proux01) [reply parsing/tok.ml]: Unfortunately, at this point (in the lexer) it is not known yet whether this token will be parsed as a number or as a user defined notation. So, if we decompose the string in a record there, we must either:
1. keep the original string along the record;
1. or be able to retrieve it from the record.

(1.) means data duplication whereas (2.) means more code, hence potentially more bugs (in particular, one is tempted to normalize the input but if "00" is lexed as "0", this can break a potential user defined "00" notation).

Thus, it seems both easier and safer to delay the spliting of the different components of the token in interp/notation.ml when we know we are actually parsing it as a number.

**Comment by proux01** [comment]
Thanks for the clarification! Indeed, there is one case that could lead to confusing type errors for the user, when she writes `t .1` expecting it to be parsed as `t` applied to `0.1` whereas it would be parsed as `fst t`. This can be fixed as either `t 0.1` or `t .10` or `t (.1)`.

Anyway, I understand the consistency point and if there is a consensus towards requiring at least one digit before the dot, I'd be happy to modify the current PR.

**Comment by maximedenes** [comment]
> Anyway, I understand the consistency point and if there is a consensus towards requiring at least one digit before the dot, I'd be happy to modify the current PR.

I think it would be good indeed.

**Comment by fajb** [comment]
> Note : Should `IZR 10200 / IZR (Z.pow_pos 10 3)` (resp. `Qmake 10200 1000`) be uninterpreted as `10200e-3` or `102e-1` or `1.02e1`?

I jump to the bandwagon...
The current encoding (using Z.pow_pos) is not  ring/micromega friendly.
Wouldn't it be possible to use the power function of the ring?
For instance, 
Goal (254e2 = 254 * 10^2)%R.
Proof.
  Fail ring.
  Fail lra.

**Comment by fajb** [comment]
Another question, somehow related.
It seems that (Z.pow_pos 10 2) cannot be changed with 100%Z when it is part of the notation.
Can you explain what I am doing wrong?
What is the way to reduce those constants?

File: parsing/tok.ml
Comment (ejgallego) [reply parsing/tok.ml]: That kind of ambiguity at the lexing level seems indeed to create many problems for little gain. What notations would we lose if we unambiguously parse numerals [thus making them not available to notations]

Comment (proux01) [reply parsing/tok.ml]: I agree but at least "0" and "1" seem commonly used.

**Comment by proux01** [comment]
> I think it would be good indeed.

Ok, done. Numerals are now [0-9]+ ([.][0-9]+)? ([eE][+-]?[0-9]+)? and `.1`, `.2` `.1414`, `.5` and `.33` are all consistenly rejected as numerals.

> The current encoding (using Z.pow_pos) is not ring/micromega friendly.

Thanks for raising this point!

> Wouldn't it be possible to use the power function of the ring?

Done, although I'm not fully satified with the solution: it yields a nat (`pow : R -> nat -> R`) that may blow up and `pow 10 2` is not convertible to `100%R` whereas `IZR (Z.pow_pos 10 2)` was.

> It seems that (Z.pow_pos 10 2) cannot be changed with 100%Z when it is part of the notation.

Interesting! I'm still trying to understand what's (not) happening here.

**Comment by fajb** [comment]
> Done, although I'm not fully satified with the solution: it yields a nat (`pow : R -> nat -> R`) that may blow up and `pow 10 2` is not convertible to `100%R` whereas `IZR (Z.pow_pos 10 2)` was.

Depending of what you do it may indeed blow up. 

I think 'ring' internally is not using 'pow' but 'pow_N' which performs iterative squaring.
There is also a powerRZ : R -> Z -> R
I won't help ring though...
For micromega, by extending the reification a bit, it should be possible to accomodate Z.pow_pos. 
I'll look into it.

**Comment by proux01** [comment]
> It seems that (Z.pow_pos 10 2) cannot be changed with 100%Z when it is part of the notation.

It appears that I was using the wrong path for `Z.pow_pos` in the OCaml module (`Coq.ZArith.BinInt` instead of `Coq.ZArith.BinIntDef`, we should probably switch to coqlib to avoid this kind of issues). Now fixed.

> I won't help ring though...
> For micromega, by extending the reification a bit, it should be possible to accomodate Z.pow_pos. I'll look into it.

It seems that a 5 line patch (c.f. https://github.com/coq/coq/pull/8764/commits/f956018eb24dcd9b0ecfee42bf278af38062816e) is enough to have ring/field accept `Z.pow_pos 10 e` as constants. So if it is simple to get them into micromega, that would be great.

Well, large exponents may still blow up when using ring (which actually computes `Z.pow_pos`) but at least not at parsing time.

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (Zimmi48) [inline doc/sphinx/user-extensions/syntax-extensions.rst]: Objective Caml -> OCaml

**Review by Zimmi48 — CHANGES_REQUESTED** [review]
You can't change the meaning of the grammar token `num` which is pervasive in the reference manual. Or else, you need to rename it and update every single place that uses it. I suggest that you introduce a new grammar token instead and find an alternative name for it.

File: doc/sphinx/user-extensions/syntax-extensions.rst
Comment (proux01) [reply doc/sphinx/user-extensions/syntax-extensions.rst]: fixed

**Comment by proux01** [comment]
> You can't change the meaning of the grammar token `num` which is pervasive in the reference manual. Or else, you need to rename it and update every single place that uses it. I suggest that you introduce a new grammar token instead and find an alternative name for it.

Indeed, sorry about that! I followed your advice.

File: parsing/tok.ml
Comment (ejgallego) [reply parsing/tok.ml]: Maybe it would less expensive still to have the order emit a num token in that case but have the notations code match it for that cases?

**Review by Zimmi48 — APPROVED** [review]

File: parsing/tok.ml
Comment (proux01) [reply parsing/tok.ml]: Sorry, I'm not sure to fully grasp your idea. Do you offer to separate numerals and notations except for "0" and "1"? How would that be "less expensive" than the current implementation?

Comment (proux01) [reply parsing/tok.ml]: @ejgallego: I tried to split the components of the numeral token in a record but it seems that coqpp is only designed to hold strings in tokens. So this would imply heavy modifications to coqpp and the result seems not worth all the trouble.

**Comment by proux01** [comment]
@fajb: I slightly modified micromega to have it accept the `IZR (Z.pow_pos _ _)` as real constants. It seems to work. Does it look good to you?

**Comment by proux01** [comment]
@fajb: c.f. https://github.com/coq/coq/pull/8764/commits/144af6f4823f027872e503b713987ec7edfb2683

**Comment by fajb** [comment]
That's ok.

**Comment by proux01** [comment]
Thanks!

**Comment by erikmd** [comment]
Hi @ejgallego, it seems the last point to discuss is related to the point you raised in [this comment](https://github.com/coq/coq/pull/8764#discussion_r251245002).
@proux01 tried to address it (cf. [his comment](https://github.com/coq/coq/pull/8764#discussion_r254707994)) but at first sight the implied refactoring would be quite cumbersome… if not impossible given the current coqpp interface… what do you think?

**Comment by ejgallego** [comment]
Oh indeed I had missed the last comment thanks Erik. Two comments:
- I think it would be less expensive in terms of code impact, tho it is not indeed clear without actually going and trying it
- coqpp should be able to be fixed, I could try to help with that but not next week

**Comment by proux01** [comment]
@ejgallego:
* I adapted coqpp (c.f. https://github.com/coq/coq/pull/8764/commits/08498f19e2b527b5600842a8670767a5f017f844, preliminary with remaining Obj.magic)
* code size is similar (c.f. https://github.com/coq/coq/pull/8764/commits/b6f5a316998a7c4f04f2ac0bdb7ed1566955f7f6, -4 lines) although we may find it a tad cleaner

What do you think? Do I remove the remaining Obj.magic?

@ppedrot You modified this code recently, so you may have an opinion here (c.f. https://github.com/coq/coq/pull/8764/commits/08498f19e2b527b5600842a8670767a5f017f844. Basically, it amounts to replacing `Plexing.pattern = string * string` with a `type 'c pattern` in `grammar.ml` (in practice instanciated with a new `'c Tok.p`)
so that
```ocaml
| Stoken : Plexing.pattern -> ('self, string) ty_symbol
```
becomes
```ocaml
| Stoken : 'c pattern -> ('self, 'c) ty_symbol
```
there.

**Comment by ppedrot** [comment]
Yes, I had a look at the code. I don't have a strong opinion on the change in type annotation, but I think it would be better to submit it as an independent PR.

**Comment by ejgallego** [comment]
Thanks @proux01 , I'll have a look ASAP.

**Comment by ppedrot** [comment]
> What do you think? Do I remove the remaining Obj.magic?

Forgot to mention: I'm fine with internal meddling with grammar.ml because it's still type unsafe, but I would personally require that you remove all the other Obj.magic in the code. It looks like you are missing typing information in there, which suggest that your patch might be slightly more invasive than what you think.

Alternatively, why do you really need the token to be an arbitrary type? You could enforce that it's a string that codes for a literal and get away with it without messing with the parser so much...

**Comment by proux01** [comment]
Thanks @ppedrot!

> I would personally require that you remove all the other Obj.magic in the code.

Done

> It looks like you are missing typing information in there, which suggest that your patch might be slightly more invasive than what you think.

Indeed, that could have had a serious impact.

> Alternatively, why do you really need the token to be an arbitrary type? You could enforce that it's a string that codes for a literal and get away with it without messing with the parser so much...

That was my first implementation but Emilio expressed a strong preference for a record (https://github.com/coq/coq/pull/8764#discussion_r251245002) so we are trying to weight the pros and cons of each solution.

> I think it would be better to submit it as an independent PR.

Sure, I'll do that if Emilio chooses this record solution.

**Comment by gares** [comment]
@proux01 what is the current status of this PR? I've another PR touching the lexer, so I'd like this to be merged first (and take the conflicts on my side, since my patch is smaller).

**Comment by proux01** [comment]
@gares we are waiting for @ejgallego to find some time to review my offers to his former criticism

I fear that make still take quite some time. So, I would say, go ahead if your PR is simpler. I'll rebase on it (won't be the first rebase of this PR anyway).

**Comment by ejgallego** [comment]
I will process this soon; together with some other PR, but anyone of @coq/parsing-maintainers could give the green light to this while I am busy.

**Comment by gares** [comment]
> I fear that may still take quite some time. So, I would say, go ahead if your PR is simpler. I'll rebase on it (won't be the first rebase of this PR anyway).

OK, thanks

**Comment by ejgallego** [comment]
I will do the review soon, however, who will assign this?

**Review by ejgallego — COMMENTED** [review]
Likely this is a naïve question, but why do you need to generalize the type of:
```
val s_token : 'c pattern -> ('self, 'c) ty_symbol
```
Couldn't it just return a regular token?

**Comment by proux01** [comment]
Not so naive, took me some time to get a precise answer.

In gramlib/grammar.mli, we have

```OCaml
    val r_next :
      ('self, 'a, 'r) ty_rule -> ('self, 'b) ty_symbol ->
        ('self, 'b -> 'a, 'r) ty_rule
    val production : ('a, 'f, Loc.t -> 'a) ty_rule * 'f -> 'a ty_production
```

where `'f` is the type of the semantic action of the rule (what's between curly braces in the mlg file). Since these actions take as arguments either `string` or `NumTok.t` (or `unit`), the type parameter of `ty_symbol` can be one of those depending on the expected token.

File: interp/constrexpr_ops.ml
Comment (ppedrot) [inline interp/constrexpr_ops.ml]: While you're at it, use `Int.equal`.

**Comment by ejgallego** [comment]
> Since these actions take as arguments either `string` or `NumTok.t` (or `unit`), the type parameter of `ty_symbol` can be one of those depending on the expected token.

So indeed couldn't these actions take a token and match against it?

**Comment by proux01** [comment]
The actions given to the `production` function are (mostly) the OCaml code appearing in the mlg files, it's the job of the `production` function to add the match against the tokens given by the lexer. Are you suggesting to change this and have to explicitly encode the match in the mlgs ? I'm not sure to follow your idea here.

**Comment by ejgallego** [comment]
> Are you suggesting to change this and have to explicitly encode the match in the mlgs ? I'm not sure to follow your idea here.

I was thinking indeed if something like that should be done, let me have a look at it myself so see how impactful would that be.

Going back to the PR, I am very sorry that my suggestion seems to have been quite misguided, I didn't anticipate the need to generalize the token production in such a way.

What are your thoughts then @proux01 @ppedrot ?

Is this approach the one to follow, was the previous one better? What should be done? Hopefully we can target a merge for next week at the latest.

**Comment by proux01** [comment]
> > Are you suggesting to change this and have to explicitly encode the match in the mlgs ? I'm not sure to follow your idea here.
> 
> I was thinking indeed if something like that should be done, let me have a look at it myself so see how impactful would that be.

I fear this would become painful to the mlg writers. But I'd be happy to be proven wrong.

> Going back to the PR, I am very sorry that my suggestion seems to have been quite misguided, I didn't anticipate the need to generalize the token production in such a way.

Well, my bad, I should have asked before going so far.

> What are your thoughts then @proux01 @ppedrot ?
> 
> Is this approach the one to follow, was the previous one better? What should be done? Hopefully we can target a merge for next week at the latest.

I have mixed feelings. In one hand, generalizing the token payload gives a somewhat neat result (other types of tokens than just strings are usable in the mlg parsers). In the other hand, it's a non negligible change to the parsing for a currently (and probably also future) limited use with only the numeral token.

In the end, I would be slightly in favor of keeping the generalized parser now that it's done but a more lightweight solution with numeral token remaining strings and being interpreted later is fine for me too (closer to my first offer after all). No strong opinion.

**Comment by ppedrot** [comment]
AFAIAC, I think that the parameterized token type is quite overengineered, but now that it has been implemented...

File: interp/constrexpr_ops.ml
Comment (proux01) [reply interp/constrexpr_ops.ml]: This is not an int but a sum type with only constant constructors. I replaced it with a proper pattern matching (https://github.com/coq/coq/pull/8764/commits/234572e65fd0039e08c9b9eec34f3124b7c2e557).

**Comment by proux01** [comment]
BTW, @ppedrot I removed some Obj.magic in grammar.ml (https://github.com/coq/coq/pull/8764/commits/69c0a9d5532478ffb6750435d6227e9eea5ca666). The only remaining ones are

```ocaml
(** FIXME: find a way to do that type-safely *)
let eq_entry : type a1 a2. a1 ty_entry -> a2 ty_entry -> (a1, a2) eq option = fun e1 e2 ->
  if (Obj.magic e1) == (Obj.magic e2) then Some (Obj.magic Refl)
  else None
```

no simple idea how to do that, and

```ocaml
let srules (type self a) ~warning (rl : a ty_production list) =
  [...]
  (* FIXME: use an universal self type to ensure well-typedness *)
  (Obj.magic (Stree t) : (self, a) ty_symbol)
```

you seem to had an idea here but I didn't catch it.

Unfortunately, I had to basically write a quadratic version of list reversal. However, I didn't observe any noticeable slowdown in Coq & stdlib compilation and indeed this function (get_symbols) is only called on very small lists (out of 818840 calls in Coq compilation, 75% of calls are made on a list of length at most 4, 91% at most 5, 99% at most 10 and the maximum is 22).

**Comment by proux01** [comment]
> ```ocaml
> let srules (type self a) ~warning (rl : a ty_production list) =
>   [...]
>   (* FIXME: use an universal self type to ensure well-typedness *)
>   (Obj.magic (Stree t) : (self, a) ty_symbol)
> ```

Removed too

**Comment by proux01** [comment]
@ejgallego so do we keep the generalized token type? In which case I'll put it in a separate PR as requested by @ppedrot (https://github.com/coq/coq/pull/8764#issuecomment-465134172).

Alternatively, we ca also defer the current PR after the feature freeze. Indeed, our main incentive for it with @erikmd is primitive floats (https://github.com/coq/coq/issues/8276) but although it's now in a good shape it is not yet ready. So not having decimal parsing in 8.10 is perfectly fine for us.

**Comment by ejgallego** [comment]
> @ejgallego so do we keep the generalized token type? In which case I'll put it in a separate PR as requested by @ppedrot ([#8764 (comment)](https://github.com/coq/coq/pull/8764#issuecomment-465134172)).

@proux01 I trust your opinion as you are way more familiar with the advantages and disadvantages of the approach than me; I'd say let's go for the solution you think is better.

I won't have the time in the short term to be very helpful with this PR here [sorry], personally, I am OK with it going in 8.10; we can also wait a bit if @ppedrot @herbelin want to comment; but I propose we don't delay the merge too much as it is going to be heavy to carry on rebasing. We can always tweak after merge.

**Comment by proux01** [comment]
> @proux01 I trust your opinion as you are way more familiar with the advantages and disadvantages of the approach than me; I'd say let's go for the solution you think is better.

Ok, let's go for the generalized token. I opened an independent PR as requested by Pierre Marie (#9815) and rebased this one over it.

**Comment by proux01** [comment]
All dependencies are merged and all discussed point have been adressed but this still lacks an assignee. @ejgallego maybe?

**Comment by ejgallego** [comment]
> @ejgallego maybe?

I'll be happy to help @proux01 , but maybe I won't be able to process this in the next days; this can be indeed important to get in 8.10 as to avoid a costly delta in terms of backports tho.

**Comment by proux01** [comment]
Rebased. CI failure for bedrock2 is definitely unrelated and failure of fcsl-pcm looks unrelated too.

**Comment by proux01** [comment]
> > All dependencies are merged and all discussed point have been adressed but this still lacks an assignee. 
> > @ejgallego maybe?
> 
> I'll be happy to help @proux01 , but maybe I won't be able to process this in the next days; this can be indeed important to get in 8.10 as to avoid a costly delta in terms of backports tho.

Anyone else among @coq/parsing-maintainers * ? (but I don't want to ask too much to @ppedrot who I already annoyed with the related #9815 )

*Looks like these alias are unfortunately unusable for non members of the github project.

**Comment by ejgallego** [comment]
> *Looks like these alias are unfortunately unusable for non members of the github project.

Can someone add @proux01 to the right group? I always forgot how to do that, maybe we should have an entry in  CONTRIBUTING "adding new contributors"? @Zimmi48

**Comment by ejgallego** [comment]
Ping @coq/parsing-maintainers as per @proux01 try above.

**Review by ejgallego — APPROVED** [review]
I am very far from an expert on numerals, but the parts of code I do understand look good; IMHO this can be merged. A few comments:
- I'd be great if people more expert than me could have a look, this includes [but not limited to] @silene , @herbelin , @JasonGross , ....
- @proux01 please add a link to the overlays Pull Request in the description.

**Comment by ejgallego** [comment]
By the way congrats on the PR @proux01 , it seems like a huge amount of work!

**Comment by Zimmi48** [comment]
> > *Looks like these alias are unfortunately unusable for non members of the github project.
> 
> Can someone add @proux01 to the right group? I always forgot how to do that, maybe we should have an entry in CONTRIBUTING "adding new contributors"? @Zimmi48

This is already there @ejgallego. I added it the last time you asked to add someone in the team: https://github.com/coq/coq/blob/master/CONTRIBUTING.md#triaging-bug-reports

**Comment by ejgallego** [comment]
> This is already there @ejgallego. I added it the last time you asked to add someone in the team: https://github.com/coq/coq/blob/master/CONTRIBUTING.md#triaging-bug-reports

Ahh sorry I promised I opened Contributing but didn't see that part. Will take of @proux01 myself.

**Review by gares — COMMENTED** [review]
I'm not an expert here but looks good to me

**Comment by proux01** [comment]
> * @proux01 please add a link to the overlays Pull Request in the description.

Done

**Comment by ejgallego** [comment]
I say let's merge this on Friday, would be great if @ppedrot @herbelin could comment or request time for review of course.

IMHO all the features of this kind should be considered "experimental " for the release they are introduced, meaning that users don't have 100% guarantee on stability in case we want to perform some refinement later on.

I suggest indeed users of this feature treat it as experimental.

**Comment by ejgallego** [comment]
Ok, I'm about to merge, I have a last question @proux01 about the changes in CoRN. Why was the overlay needed, should we add some more migration tips to the CHANGES file?

IMHO if this change is going to break user developments in any way we should try to document the change of semantics better, the current text seems a bit too concise.

**Comment by proux01** [comment]
> I have a last question @proux01 about the changes in CoRN. Why was the overlay needed,

CORN did introduce custom notations on the numeral tokens 2, 3 and 4 which were conflicting with the newly introduced one in Q.

> should we add some more migration tips to the CHANGES file?

That seems like a rare case (only happened in CORN in the CI) but you're right, I just updated CHANGES.md with that point and the experimental status you suggested in your previous comment.

Let me know in case a rebase would be needed.

**Comment by ejgallego** [comment]
Thanks @proux01 , I will merge as soon as the CI passes. I hope we don't need a rebased.

**Comment by proux01** [comment]
All overlays are merged so we are all set.

Thanks @ejgallego and thanks to everyone who reviewed or contributed in the discussion!

**Pull request merged**
